R/defs.bzl

```diff
@@ -711,3 +715,128 @@ def r_package_with_test(pkg_name, pkg_srcs, pkg_deps, pkg_suggested_deps=[], tes
         pkg = pkg_name,
         suggested_deps = pkg_suggested_deps,
     )
+
+load(":runfiles_commands.bzl", "runfiles_commands")
+
+def _r_binary_impl(ctx):
+    if len(ctx.files.srcs) > 1:
+        fail("only one srcs file is allowed; got " + len(ctx.files.srcs))
+
+    ldeps = _library_deps(ctx.attr.deps, path_prefix=ctx.workspace_name + "/")
+
+    tools = _executables(ctx.attr.tools)
+    for dep in ctx.attr.deps:
+        tools += dep[RPackage].transitive_tools
+
+    script = "\n".join([
+        "#!/bin/bash",
+        "set -euo pipefail",
+        "",
+    ] + _env_vars(ctx.attr.env_vars) + runfiles_commands() + [("\n".join([
+        "",
+        "export PATH=\"${{PATH}}:{0}\"",
+        "",
+        "export R_LIBS_USER=$(mktemp -d)",
+        "pushd ${{RUNFILES}} > /dev/null",
+        ldeps["symlinked_library_command"],
+        "",
+        "pushd {2} > /dev/null",
+        "cleanup() {{",
+        "  set -e",
+        "  popd > /dev/null",
+        "  rm -rf ${{R_LIBS_USER}}",
+        "}}",
+        "",
+        "set +e",
+        _Rscript + "{3} {1} \"$@\"",
+        "code=$?",
+        "if [ $code -ne 0 ]; then",
+        "  cleanup",
+        "  echo \"Rscript error code: $code\"",
+        "  exit $code",
+        "fi",
+        "",
+        "cleanup",
+    ])).format(
+        _path(tools), ctx.files.srcs[0].short_path, ctx.workspace_name,
+        ctx.attr.rscript_args)])
+
+    ctx.actions.write(output = ctx.outputs.executable, content = script)
+
+    runfiles = ctx.runfiles(
+        files = ldeps["lib_files"] + ctx.files.srcs + ctx.files.data,
+        transitive_files = tools)
+    return [DefaultInfo(runfiles = runfiles)]
+
+r_binary = rule(
+    attrs = {
+        "srcs": attr.label_list(
+            # https://cran.r-project.org/doc/manuals/r-release/R-exts.html
+            allow_files = FileType([
```

Comment:
FileType is actually deprecated and unnecessary now. You can simply say `allow_single_file = [".R", ".r", ".S", ".s", ".q"]`, etc.

Reply:
I am actually thinking about allowing any kind of executable file in here, does not have to be run through the Rscript interpreter. This rule will simply set up the environment for that executable to run (R_LIBS_USER, etc.).

This grants more flexibility because users will be able to also use their shell scripts as src for r_test; the script can rely on bazel to install the declared package dependencies in the user's R library.

If we do this, then Rscript args will need to be in the shebang line of the src file, and can not be an attribute of the rule.

Reply:
I haven't thought of that, I like this idea as well.

Reply:
However, I think that if we want to keep in line with, e.g., the various rules in docker_rule (more on that for a r_image rule I'm updating), we should check if there is a shebang, and if there is none, just use Rscript.